lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+
+from ansible.module_utils.urls import open_url
+
+
+display = Display()
+
+MANIFEST_FORMAT = 1
+
+
+class CollectionRequirement(object):
+
+    _FILE_MAPPING = [('MANIFEST.json', 'manifest'), ('FILES.json', 'files')]
+
+    def __init__(self, namespace, name, path, source, versions, requirement, force, parent=None, validate_certs=True,
+                 metadata=None, files=None, skip=False):
+        """
+        Represents a collection requirement, the versions that are available to be installed as well as any
+        dependencies the collection has.
+
+        :param namespace: The collection namespace.
+        :param name: The collection name.
+        :param path: The path to the collection tarball if it has already been downloaded.
+        :param source: The Galaxy server URL to download if the collection from Galaxy.
+        :param versions: A list of versions of the collection that is available
+        :param requirement: The version requirement string used to verify the list of versions fit the requirements.
+        :param force: Whether the force flag applied to the collection.
+        :param parent: The name of the parent the collection is a dependency of.
+        :param validate_certs: Whether to validate the Galaxy server certificate.
+        :param metadata: The collection metadata dict if it has already been retrieved.
+        :param files: The files that exist inside the collection. This is based on the FILES.json file inside the
+            collection artifact.
+        :param skip: Whether to skip installing the collection, should be set if the collection is already installed
+            and force is not set.
+        """
+        self.namespace = namespace
+        self.name = name
+        self.path = path
+        self.source = source
+        self.versions = set(versions)
+        self.force = force
+        self.skip = skip
+        self.required_by = []
+        self._validate_certs = validate_certs
+
+        self._metadata = metadata
+        self._files = files
+        self._galaxy_info = None
+
+        self.add_requirement(parent, requirement)
+
+    def __str__(self):
+        return "%s.%s" % (self.namespace, self.name)
+
+    @property
+    def latest_version(self):
+        versions = [v for v in self.versions if v != '*']
+        if versions:
+            versions = sorted(versions, key=LooseVersion)
+            return versions[-1]
+        else:
+            return '*'
+
+    @property
+    def dependencies(self):
+        if self._metadata:
+            return self._metadata['dependencies']
+        elif len(self.versions) > 1:
+            return None
+
+        self._get_metadata()
+        return self._metadata['dependencies']
+
+    def add_requirement(self, parent, requirement):
+        self.required_by.append((parent, requirement))
+        new_versions = set([v for v in self.versions if self._meets_requirements(v, requirement)])
```

Comment:
eliminate intermediate list
```suggestion
        new_versions = set(v for v in self.versions if self._meets_requirements(v, requirement))
```